rust/agama-server/src/software/web.rs

```diff
@@ -0,0 +1,242 @@
+))]
+async fn products(
+    State(state): State<SoftwareState<'_>>,
+) -> Result<Json<Vec<Product>>, SoftwareError> {
+    let products = state.product.products().await?;
+    Ok(Json(products))
+}
+
+/// Represents a pattern.
+///
+/// It augments the information coming from the D-Bus client.
+#[derive(Serialize, utoipa::ToSchema)]
+pub struct PatternEntry {
+    #[serde(flatten)]
+    pattern: Pattern,
+    selected_by: SelectedBy,
+}
+
+/// Returns the list of software patterns.
+///
+/// * `state`: service state.
+#[utoipa::path(get, path = "/software/patterns", responses(
+    (status = 200, description = "List of known software patterns")
+))]
+async fn patterns(
+    State(state): State<SoftwareState<'_>>,
+) -> Result<Json<Vec<PatternEntry>>, SoftwareError> {
+    let patterns = state.software.patterns(true).await?;
+    let selected = state.software.selected_patterns().await?;
+    let items = patterns
+        .into_iter()
+        .map(|pattern| {
+            let selected_by: SelectedBy = selected
+                .get(&pattern.id)
+                .copied()
+                .unwrap_or(SelectedBy::None);
+            PatternEntry {
+                pattern,
+                selected_by,
+            }
+        })
+        .collect();
+
+    Ok(Json(items))
+}
+
+/// Sets the software configuration.
+///
+/// * `state`: service state.
+/// * `config`: software configuration.
+#[utoipa::path(put, path = "/software/config", responses(
+    (status = 200, description = "Set the software configuration")
+))]
+async fn set_config(
+    State(state): State<SoftwareState<'_>>,
+    Json(config): Json<SoftwareConfig>,
+) -> Result<(), SoftwareError> {
+    if let Some(product) = config.product {
+        state.product.select_product(&product).await?;
+    }
+
+    if let Some(patterns) = config.patterns {
+        state.software.select_patterns(&patterns).await?;
+    }
+
+    Ok(())
+}
+
+/// Returns the software configuration.
+///
+/// * `state` : service state.
+#[utoipa::path(get, path = "/software/config", responses(
+    (status = 200, description = "Software configuration")
+))]
+async fn get_config(
+    State(state): State<SoftwareState<'_>>,
+) -> Result<Json<SoftwareConfig>, SoftwareError> {
+    let product = state.product.product().await?;
+    let patterns = state.software.user_selected_patterns().await?;
+    let config = SoftwareConfig {
+        patterns: Some(patterns),
+        product: Some(product),
+    };
+    Ok(Json(config))
+}
+
+#[derive(Serialize, utoipa::ToSchema)]
+/// Software proposal information.
+struct SoftwareProposal {
+    /// Space required for installation. It is returned as a formatted string which includes
+    /// a number and a unit (e.g., "GiB").
+    size: String,
+}
+
+/// Returns the proposal information.
+///
+/// At this point, only the required space is reported.
+#[utoipa::path(
+    get, path = "/software/proposal", responses(
+        (status = 200, description = "Software proposal")
```

Comment:
does it also somehow document how returning json will look like? Will it link it with SoftwareProposal struct automatic?

Reply:
No, it does not :-(